Implement a Docker image generator gradle plugin for Spark applications

mccheah opened the pull request:
Allows specifying the jars for the application's classpath as well as a custom base image.

robert3005 commented on .circleci/config.yml:
```diff
@@ -479,6 +495,10 @@ workflows:
           requires:
             - build-sbt
           <<: *all-branches-and-tags
+      - run-spark-docker-gradle-plugin-tests:
```
you need to wire publishing as well

mccheah replied:
Should be included because we put the gradle publish call in `publish_functions.sh`

robert3005 commented on dev/publish_functions.sh:
```diff
@@ -22,6 +22,9 @@ publish_artifacts() {
   echo "</server></servers></settings>" >> $tmp_settings
 
   ./build/mvn -T 1C --settings $tmp_settings -DskipTests "${PALANTIR_FLAGS[@]}" deploy
+  cd resource-managers/kubernetes/docker
```
Do you actually need to cd here (would be less error prone if you don't need to)?

mccheah replied:
Yep, see https://github.com/palantir/spark/pull/381/files/5a847bbc318d99e5c89e5e61a07fc7b2f2ce1dbc#r196244780

robert3005 commented on dev/build-spark-docker-gradle-plugin.sh:
```diff
@@ -0,0 +1,24 @@
+set -euo pipefail
+ROOT=$(git rev-parse --show-toplevel)
+cd "$ROOT/resource-managers/kubernetes/docker"
+./gradlew --info compileJava compileTestJava
```
ideally you can just call `resource-managers/kubernetes/docker/gradlew`?

robert3005 replied:
why not `build`?

mccheah replied:
Can resolve with https://stackoverflow.com/questions/15374409/how-to-pass-build-gradle-directory-to-gradlew

robert3005 commented:
@mccheah looks good - mostly would like to figure out if we can remove some boilerplate gradle scripts

robert3005 commented on resource-managers/kubernetes/docker/settings.gradle:
```diff
@@ -0,0 +1,18 @@
+ * limitations under the License.
+ */
+
+rootProject.name = 'spark-docker-image-generator_2.11'
```
don't add 2.11 - this isn't a scala project?

robert3005 commented on resource-managers/kubernetes/docker/src/main/java/org/apache/spark/deploy/kubernetes/docker/gradle/DeployScriptsTask.java:
```diff
@@ -0,0 +1,124 @@
+             new TarArchiveInputStream(dockerResourcesGZipped)) {
+      TarArchiveEntry nextEntry = dockerResourcesTar.getNextTarEntry();
+      while (nextEntry != null) {
+        boolean extracted = maybeCopyTarEntry(dockerResourcesTar,
```
can we extract once and copy?

dansanduleac replied:
This whole function can be described using a `Sync` task (or, if we need `dockerBuildDir` to be defined lazily, then can use an inline action `project.sync(spec -> ...)`.
Picking selective files from a tar is easy, though there is a bit more logic necessary if you wanna rename paths... as an example to get started:
```java
// Not 100% sure if you can give it a URL but from looking into the code it should work
// Alternatively you can implement a `ReadableResource` and pass that to tarTree
FileTree tree = project.tarTree(getClass().getResource("/docker-resources.tgz"));
project.sync(spec -> {
    spec.from(tree, s2 -> {
        s2.include("docker-resources/bin", "docker-resources/sbin");
    });
    // TODO implement stripFirstDirectoryFrom
    spec.eachFile(fcd -> fcd.setRelativePath(stripFirstDirectoryFrom(fcd.getRelativePath())));
    spec.into(outputDir);
});
```

dansanduleac replied:
Then you don't need `deleteAndMkDirs` either, Sync automatcally copies just what was included by the `from` blocks into the `into` directory, removing everything else that's not supposed to be there.

robert3005 commented on resource-managers/kubernetes/docker/src/main/java/org/apache/spark/deploy/kubernetes/docker/gradle/GenerateDockerFileTask.java:
```diff
@@ -0,0 +1,109 @@
+                        return String.format("FROM %s", getBaseImage());
+                      } else if (line.equals(
+                          "    apk add --no-cache bash tini libc6-compat && \\")) {
+                        return line.replace("libc6-compat ", "");
```
why not just remove this from image?

mccheah replied:
Good question, we'll discuss this in the community - probably we want custom image authors to be providing this component themselves. For now I think it's fine to fork the Dockerfile.

robert3005 replied:
Wait - this is our fork, let's just remove it from dockerfile here

robert3005 commented on resource-managers/kubernetes/docker/src/main/java/org/apache/spark/deploy/kubernetes/docker/gradle/GenerateDockerFileTask.java:
```diff
@@ -0,0 +1,109 @@
+               BufferedReader sourceDockerFileBuffered =
+                   new BufferedReader(sourceDockerFileReader)) {
+            List<String> fileLines = Collections.unmodifiableList(
+                sourceDockerFileBuffered.lines()
```
Files.lines()? Just remember to close the stream

mccheah replied:
The source docker file here is embedded inside the tarball. You can't get a `Path` or a `File` object directly from the archive entry, at least not one which can be used to open a stream.

robert3005 replied:
How about not embedding a tgz? Not sure how many files are actually there. Then you could use https://google.github.io/guava/releases/21.0/api/docs/com/google/common/io/Resources.html#readLines-java.net.URL-java.nio.charset.Charset-

robert3005 commented on resource-managers/kubernetes/docker/src/main/java/org/apache/spark/deploy/kubernetes/docker/gradle/SparkDockerPlugin.java:
```diff
@@ -0,0 +1,144 @@
+          + dockerBuildDirectory.getAbsolutePath());
+    }
+
+    project.afterEvaluate(evaluatedProject -> {
```
This shouldn't be using afterEvaluate - dependencies should be setup via task inputs and outputs

mccheah replied:
Some of the wiring here relies on the `jar` task being set up already -n are we saying that the plugin user should be passing that output jar in specifically? I'm not sure how we get the output of the `jar` task injected into the image without using `afterEvaluate` to verify that the `jar` task exists in the first place.

dansanduleac replied:
The jar task comes from `JavaPlugin`, you can do smth like 
```
project.getPluginManager().withPlugin("java", javaPlugin -> {
    // access the jar task here, knowing it has been created
});
```

mccheah commented:
@dansanduleac do you have any advice for how to create tasks given the setup of the extension, without using the afterEvaluate method?

I looked at https://github.com/palantir/gradle-docker/blob/develop/src/main/groovy/com/palantir/gradle/docker/PalantirDockerPlugin.groovy#L117 and in particular it generated a separate docker tag task for every tag to be generated. But it had to use `afterEvaluate` to do that, and I'm wondering if we can do something more idiomatic.

Taking a step back though I wouldn't be entirely opposed to the idea of only allowing for a single tag per Spark application image, so if accomplishing this would require an overly complex setup we can make this only assign one tag.

mccheah commented:
Hm, we could define a task `Rule` which checks if the extension defined the given tag and constructs the task on the fly. Would that work? https://www.oreilly.com/library/view/gradle-beyond-the/9781449373801/ch03.html

mccheah commented:
Actually `Rule` and Model based Gradle development is a pretty significant paradigm shift altogether. It might be correct for this situation though. Will do some experimentation with the APIs.

dansanduleac commented:
Yep if you want to create a different number of tasks based on runtime values, then you need afterEvaluate (and wire them up via dependencies to a task that you explicitly create outside of afterEvaluate). But perhaps you can do _only_ that bit in afterEvaluate, that would simplify the flow a lot.
Not sure if it's worth complicating matters with the `Rule` api which seems to me as powerful as scala implicits but with the added blackboxiness of doing everything via reflection.

robert3005 commented:
@mccheah how about instead of putting docker files as a tgz in the jar we publish them separately? This would be akin to how gradle-baseline does things where configuration is an input to the plugin. This would let us remove some of the weird file reading and in general be more gradle friendly

mccheah commented:
Ok, @yifeih @dansanduleac @robert3005 I picked this up again. Addressed all comments - using lazy properties as much as possible. One after evaluate block is used to create new tasks based on the extension's configuration in the user's project. Please take a look.

mccheah commented:
@robert3005 @dansanduleac comments are addressed and the build is green. Can we take another look at this?

mccheah commented:
@dansanduleac @robert3005 ping on this

robert3005 commented on dev/run-spark-docker-gradle-plugin-tests.sh:
```diff
@@ -0,0 +1,22 @@
+set -euo pipefail
+ROOT=$(git rev-parse --show-toplevel)
+DOCKER_PLUGIN_PROJECT_DIR=$ROOT/resource-managers/kubernetes/docker
+$DOCKER_PLUGIN_PROJECT_DIR/gradlew -p $DOCKER_PLUGIN_PROJECT_DIR --info test
```
add gradle to root and only define one subproject with that path so we can just run `./gradlew test` don't see any benefit nesting it so far and keeping it constrained

robert3005 commented on resource-managers/kubernetes/docker/build.gradle:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  dependencies {
+    classpath 'gradle.plugin.com.palantir:gradle-circle-style:1.1.2'
```
can you use baseline?

robert3005 commented on resource-managers/kubernetes/docker/build.gradle:
```diff
@@ -0,0 +1,109 @@
+}
+
+dependencies {
+  compileOnly gradleApi()
```
make it compile - this is a weird dependency

robert3005 commented on resource-managers/kubernetes/docker/build.gradle:
```diff
@@ -0,0 +1,109 @@
+  testCompile gradleTestKit()
+}
+
+def getGitRepoRoot = { ->
```
this shouldn't be necessary - let's make root project be in the root of the repo

robert3005 commented on resource-managers/kubernetes/docker/gradle/publish.gradle:
```diff
@@ -0,0 +1,56 @@
+ */
+
+apply plugin: 'com.jfrog.bintray'
+apply plugin: 'nebula.maven-base-publish'
```
use `nebula.maven-publish` and remove line below as well as bintrayupload dependencies and publications. Just make publish depend on bintrayupload (see baseline for reference)

robert3005 commented on resource-managers/kubernetes/docker/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -0,0 +1,22 @@
+
+distributionBase=GRADLE_USER_HOME
+distributionPath=wrapper/dists
+distributionUrl=https\://services.gradle.org/distributions/gradle-4.7-bin.zip
```
latest

robert3005 commented on resource-managers/kubernetes/docker/src/main/dockerfiles/spark/Dockerfile:
```diff
@@ -41,8 +41,8 @@ COPY ${spark_jars} /opt/spark/jars
 COPY ${img_path}/spark/entrypoint.sh /opt/
-COPY examples /opt/spark/examples
-COPY data /opt/spark/data
+#COPY examples /opt/spark/examples
```
remove instead of commenting

robert3005 commented on resource-managers/kubernetes/docker/src/main/java/org/apache/spark/deploy/kubernetes/docker/gradle/SparkDockerPlugin.java:
```diff
@@ -0,0 +1,180 @@
+          dockerBuildTask.dependsOn("sparkDockerPrepare");
+        });
+    Task tagAllTask = project.getTasks().create("sparkDockerTag");
+    LazyExecTask pushAllTask = project.getTasks().create(
```
aren't tasks lazy anywa?

mccheah replied:
You need to lazily evaluate the extension but `Exec` only allows passing the String values up-front.

mccheah replied:
Discussed in https://github.com/palantir/spark/pull/381#discussion_r196244415

robert3005 commented on resource-managers/kubernetes/docker/src/main/java/org/apache/spark/deploy/kubernetes/docker/gradle/SparkDockerPlugin.java:
```diff
@@ -0,0 +1,180 @@
+          }));
+          task.setCommandLine(pushCommandLine);
+        });
+    project.afterEvaluate(evaluatedProject -> {
```
why does this require afterevaluate?

mccheah replied:
Discussed in https://github.com/palantir/spark/pull/381#issuecomment-399412708

robert3005 commented on resource-managers/kubernetes/docker/src/main/java/org/apache/spark/deploy/kubernetes/docker/gradle/SparkDockerPlugin.java:
```diff
@@ -0,0 +1,180 @@
+      } else {
+        sparkDockerRuntimeConfiguration.extendsFrom(runtimeConfiguration);
+      }
+      Task jarTask = project.getTasks().getByName("jar");
```
upgrade to 4.9 and use versions from https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#sec:old_vs_new_configuration_api_overview

robert3005 commented on resource-managers/kubernetes/docker/src/main/java/org/apache/spark/deploy/kubernetes/docker/gradle/SparkDockerExtension.java:
```diff
@@ -0,0 +1,68 @@
+
+public class SparkDockerExtension {
+
+  private String baseImage = "openjdk:8-alpine";
```
properties please

robert3005 commented on resource-managers/kubernetes/docker/src/main/java/org/apache/spark/deploy/kubernetes/docker/gradle/SparkDockerPlugin.java:
```diff
@@ -0,0 +1,180 @@
+      Project project,
+      SparkDockerExtension extension) {
+    Property<String> imageNameProperty = project.getObjects().property(String.class);
+    imageNameProperty.set(project.provider(extension::getImageName));
```
once you make extension have properties you should be able to avoid those two lines

robert3005 commented on resource-managers/kubernetes/docker/src/test/java/org/apache/spark/deploy/kubernetes/docker/gradle/DockerBuildTaskSuite.java:
```diff
@@ -0,0 +1,84 @@
+        dockerBuildDirectory.getAbsolutePath());
+  }
+
+}
```
missing newline character

robert3005 commented on resource-managers/kubernetes/docker/src/test/java/org/apache/spark/deploy/kubernetes/docker/gradle/DockerBuildTaskSuite.java:
```diff
@@ -0,0 +1,84 @@
+import java.io.File;
+import java.io.IOException;
+
+public final class DockerBuildTaskSuite {
```
prefer use of mockitojunitrunner since `initMocks` no longer works in junit5 and it's more obvious to migrate since the runner no longer exists there

mccheah commented:
Agreed that properties should be used in the extension. We can move the project up to the root directory, I mostly didn't want to make the root directory too different from stock Spark's (i.e. isolate the added code to be in the resource-managers/kubernetes/docker subdirectory). The use of `afterEvaluate` as well as custom exec tasks is a necessary evil - please refer to earlier discussion in the PR.

robert3005 commented:
Makes sense - one more question though. Why do we need tags? They're not useful and it's easier to just publish snapshot and release versions. Internally we stopped relying on latest and snapshot long time ago and I don't see a good reason why you'd want them.

mccheah commented:
The cue is mainly taken from gradle-docker: https://github.com/palantir/gradle-docker/blob/develop/src/main/groovy/com/palantir/gradle/docker/DockerExtension.groovy#L35. But this is more by convention  and consistency with that plugin. In practice tag should more or less be the git tag, or the git hash if it's a snapshot..

mccheah commented on .baseline/checkstyle/checkstyle-suppressions.xml:
```diff
@@ -0,0 +1,26 @@
+
+    <!-- Generated code should not be subjected to checkstyle. -->
+    <suppress files="[/\\].*[/\\]generated.*[/\\]" checks="." />
+    <suppress files="[/\\].*[/\\]docker-resources.*[/\\]" checks="." />
```
Avoid invoking checkstyle on shell scripts that are placed in the resources folder

mccheah commented on .baseline/checkstyle/checkstyle.xml:
```diff
@@ -0,0 +1,446 @@
+    </module>
+    <module name="Header">
+        <!--<property name="header" value="^/\*$\n^ \* Licensed to the Apache Software Foundation (ASF) under one or more$\n"/>-->
+        <property name="headerFile" value="${config_loc}/header.txt"/>
```
Make the header use only the Apache license

mccheah commented on .baseline/checkstyle/checkstyle.xml:
```diff
@@ -0,0 +1,446 @@
+        <module name="Indentation"> <!-- Java Style Guide: Block indentation: +2 spaces -->
+            <property name="arrayInitIndent" value="4"/>
+            <property name="lineWrappingIndentation" value="4"/>
+            <property name="basicOffset" value="2"/>
```
All indentation is 2 spaces to be consistent with the rest of the Spark codebase

mccheah commented on .baseline/checkstyle/checkstyle.xml:
```diff
@@ -0,0 +1,446 @@
+            <message key="illegal.type" value="Do not use Guava caches, they are outperformed by and harder to use than Caffeine caches"/>
+        </module>
+        <!-- Use Spark's built-in import ordering -->
+        <module name="ImportOrder">
```
Import ordering matches the rest of Spark, although from what we can tell this hasn't been enforced consistently. https://spark.apache.org/contributing.html

mccheah commented on dev/.rat-excludes:
```diff
@@ -113,3 +113,8 @@ structured-streaming/*
 kafka-source-initial-offset-version-2.1.0.bin
 kafka-source-initial-offset-future-version.bin
 vote.tmpl
+gradle-build/*
```
Avoid inspecting files built by Gradle for license checks

mccheah commented:
@robert3005 addressed comments. I've noted the differences to stock baseline in comments here.

if you feel strongly about not supporting multiple tags then we can remove it.

mccheah commented on build.gradle:
```diff
@@ -0,0 +1,64 @@
+allprojects {
+  group 'org.apache.spark'
+  version System.env.CIRCLE_TAG ?: gitVersion()
+  buildDir = 'gradle-build'
```
Because Spark's root directory already contains a folder called `build`, presumably for sbt, and we can't delete its contents freely with `./gradlew clean`.

mccheah commented on resource-managers/kubernetes/docker/build.gradle:
```diff
@@ -0,0 +1,109 @@
+    into 'kubernetes/dockerfiles/spark'
+  }
+
+  from(fileTree("$getGitRepoRoot/bin")) {
```
I believe this doesn't apply anymore, given how we've rearranged the Gradle project.

mccheah replied:
We instead use `rootDir` instead of `getGitRepoRoot`. if we were to use relative paths, we'd have to refer to the parent path, e.g. `from(fileTree("../bin"))`. I'm generally against referencing the parent path this way because it requires one to know where the script is running from. Referencing the absolute path gives us a path that's consistent regardless of where the script is running from or if we modify the directory structure of this project later.

mccheah commented on resource-managers/kubernetes/docker/build.gradle:
```diff
@@ -0,0 +1,109 @@
+  testCompile gradleTestKit()
+}
+
+def getGitRepoRoot = { ->
```
This function isn't in the latest version.

robert3005 commented on dev/run-spark-docker-gradle-plugin-tests.sh:
```diff
@@ -0,0 +1,20 @@
+# limitations under the License.
+#
+set -euo pipefail
+./gradlew --info --stacktrace test
```
we shouldn't need this - just call the gradle command, you gain nothing by wrapping it

robert3005 commented:
@mccheah lets remove the wrapping scripts after that it's LGTM from me

robert3005 commented on spark-docker-image-generator/src/main/java/org/apache/spark/deploy/kubernetes/docker/gradle/SparkDockerPlugin.java:
```diff
@@ -0,0 +1,178 @@
+          commandLine.add(extension.getImageName());
+          task.setCommandLine(commandLine);
+        });
+    project.afterEvaluate(evaluatedProject -> {
```
I think this is not necessary as long as we make configuration of the extension lazy. If you look at how MavenPublish plugin works from 4.8 onwards is that it creates tasks in `all` action on publication container. Hence, we can assume that any time you'd need an after evaluate is an error since it means the user force project evaluation at some point.

Simple thing to do here would be to make tags `NamedDomainObjectContainer` and then you can lazily register things by name which would be the tag.

robert3005 replied:
This can be done separately as long as it's done

mccheah commented:
@dansanduleac @robert3005 going to merge after latest CI passes if there are no more comments.

robert3005 commented:
Cool with me, just remember to squash

mccheah commented:
Whoops it broke master - will fix
